src/irr/asset/CImageLoaderTGA.cpp

```diff
@@ -9,12 +9,89 @@
 #include "IReadFile.h"
 #include "os.h"
 #include "irr/asset/format/convertColor.h"
-#include "irr/asset/ICPUImageView.h"
+#include "irr/asset/ICPUImage.h"
 
 namespace irr
 {
 namespace asset
 {
+	/*
+		For color pallete stream.
+		Create an image containing a single row from taking an ICPUBuffer
+		as a single row and convert it to any format. Since it's
+		data may not only limit to stuff being displayed on a screen,
+		there is an optiomal parameter for bufferRowLength pitch that
+		is helpful while dealing with specific data which needs it.
+	*/
+
+	template<E_FORMAT inputFormat, E_FORMAT outputFormat>
+	static inline core::smart_refctd_ptr<ICPUImage> createSingleRowImageFromRawData(core::smart_refctd_ptr<asset::ICPUBuffer> inputBuffer, bool createWithBufferRowLengthPitch = false)
+	{
+		auto rowData = inputBuffer->getPointer();
+		const uint32_t texelOrBlockLength = inputBuffer->getSize() / asset::getTexelOrBlockBytesize(inputFormat);
+
+		using CONVERSION_FILTER = CConvertFormatImageFilter<inputFormat, outputFormat>;
+		CONVERSION_FILTER convertFilter;
+		CONVERSION_FILTER::state_type state;
+
+		auto createImage = [&](E_FORMAT format, bool copyInputMemory = true)
+		{
+			const auto texelOrBlockByteSize = asset::getTexelOrBlockBytesize(format);
+			const uint32_t pitchTexelOrBlockLength = createWithBufferRowLengthPitch ? asset::IImageAssetHandlerBase::calcPitchInBlocks(texelOrBlockLength, texelOrBlockByteSize) : texelOrBlockLength;
+
+			ICPUImage::SCreationParams imgInfo;
+			imgInfo.format = format;
+			imgInfo.type = ICPUImage::ET_1D;
+			imgInfo.extent = { texelOrBlockLength * asset::getBlockDimensions(format).X, 1, 1 };
+			imgInfo.mipLevels = 1u;
+			imgInfo.arrayLayers = 1u;
+			imgInfo.samples = ICPUImage::ESCF_1_BIT;
+			imgInfo.flags = static_cast<IImage::E_CREATE_FLAGS>(0u);
+
+			auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1u);
+			auto texelBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(texelOrBlockByteSize * pitchTexelOrBlockLength);
+
+			if (copyInputMemory)
+				texelBuffer = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(texelOrBlockByteSize * texelOrBlockLength, rowData, core::adopt_memory);
+
+			ICPUImage::SBufferCopy& region = regions->front();
+
+			region.imageSubresource.mipLevel = 0u;
+			region.imageSubresource.baseArrayLayer = 0u;
+			region.imageSubresource.layerCount = 1u;
+			region.bufferOffset = 0u;
```

Comment:
this function is overly complex, I'd just make a function that takes an input `ICPUBuffer` and the same input/output format template parameters, and create an output `ICPUBuffer`

You can still iterate over memory and call 
```cpp
const void* srcPix[] = {reinterpret_cast<const uint8_t*>(inputBuffer->getPointer())+texel*inputBlockByteSize,nullptr,nullptr,nullptr};
convertColor<inputFormat,outputFormat>(srcPix,reinterpret_cast<const uint8_t*>(outputBuffer->getPointer())+texel*outputBlockByteSize,0,0);
```
on each texel

the whole function will be 10 lines instead of 60

Reply:
fixed in upcoming commit